src/test_typing_extensions.py

```diff
@@ -8098,6 +8100,563 @@ def test_pep_695_generics_with_future_annotations_nested_in_function(self):
+                # locals
+                (outer_locals, None),
+            ):
+                owner = None
+                # Argument for ForwardRef
+                if isinstance(annot, type):
+                    annot = annot.__name__
+                else:
+                    annot = str(annot)
+                # 3 cases for the different formats
+                if isinstance(expected, dict):
+                    case = _unroll_subcase(
+                        annot,
+                        expected,
+                        format,
+                        is_argument,
+                        is_class,
+                        globalns,
+                        localns,
+                    )
+                    yield case
+                # Multiple cases depending on other parameters
+                elif type(expected) is list:  # noqa: E721 # avoid _ConcatenateGenericAlias
+                    yield from filter(
+                        None,
+                        (
+                            _unroll_subcase(
+                                annot,
+                                sub_expected,
+                                format,
+                                is_argument,
+                                is_class,
+                                globalns,
+                                localns,
+                            )
+                            for sub_expected in expected
+                        ),
+                    )
+                # Single value
+                else:
+                    # Change expected to TypeError if it will fail typing._type_check
+                    if format != Format.STRING and not inspect.isclass(expected):
+                        invalid_generic_forms = (Generic, Protocol)
+                        if is_class is False:
+                            invalid_generic_forms += (ClassVar,)
+                            if is_argument:
+                                invalid_generic_forms += (Final,)
+                        elif not _FORWARD_REF_HAS_CLASS and is_argument:
+                            invalid_generic_forms += (ClassVar, Final,)
+                        if get_origin(expected) in invalid_generic_forms:
+                            expected = TypeError
+                    type_params = None
+                    yield (
+                        annot,
+                        expected,
+                        format,
+                        is_argument,
+                        is_class,
+                        globalns,
+                        localns,
+                        type_params,
+                        owner,
+                    )
+        # Test cases
+        for (
+            annot,
+            expected,
+            format,
+            is_argument,
+            is_class,
+            globalns,
+            localns,
+            type_params,
+            owner,
+        ) in unroll_cases(cases, locals()):
```

Comment:
I'm not sure all this indirection makes the test cases better. It becomes harder to verify the tests behave as expected with so much logic between the definition of the tests and the actual tests.

Reply:
I'd simply write explicit test cases for some important cases, instead of attempting to create an abstraction around them.

Reply:
Done, Yes I think that looks better now. Thank you.